### PR TITLE
[runtime] add test for receiving output from a wasm module

### PR DESCRIPTION
**Pull request by Officeyutong**
Completed test `test_receive_wasm_bpf_module_output`. 

- I use the epoch interruption to implement a more precise time monitoring https://docs.rs/wasmtime/latest/wasmtime/struct.Store.html#method.set_epoch_deadline
- When the time exceeded, the wasm program will be terminated with a trap. At this time the control flow will be back to the rust code, so we can examine the content of `WritePipe`s.
- A seperate thread is used to call the wasm function. The original main thread is used to increase epoch when the timelimit exceeded.
- To allow calling wasm program in another thread, I splitted the original `run_wasm_bpf_module` function. Keep `engine` away from `function` and `store`. The first one is the only things needed to increase epoch, and the last two ones are the only thing needed to call the entry function of a wasm program. So just move the function instance and store to the new thread and run the wasm program. The main thread goes to sleep and increase epoch

### CODE REVIEW

File: runtime/wasm-bpf-rs/src/state.rs [hunk 1]
```diff
@@ -39,6 +39,10 @@ pub struct AppState {
     pub callback_func_name: String,
     pub wrapper_called: bool,
 }
+
+// We need a seprate thread to monitor the running process of a wasm program, so we must ensure that AppState to be Send
+#[cfg(test)]
+unsafe impl Send for AppState {}
```
Comment (yunwei37) [inline runtime/wasm-bpf-rs/src/state.rs:45]: Why this cannot be safe?

File: runtime/wasm-bpf-rs/src/lib.rs [hunk 2]
```diff
@@ -145,18 +205,31 @@ mod tests {
...
+            // There will be an epoch interruption in the wasm program.
+            engine.increment_epoch();
         });
-        thread::sleep(std::time::Duration::from_secs(timeout_sec));
-        // kill the thread
+        let result = func.run();
+        if let Err(e) = result {
+            // We can't distinguish epoch trap and other errors easily...
+            println!("{}", e.to_string());
+        }
```
Comment (yunwei37) [inline runtime/wasm-bpf-rs/src/lib.rs:232]: can the wasm module continue to run or recover from the trap?

File: runtime/wasm-bpf-rs/src/state.rs (hunk 1 above)
Comment (yunwei37) [reply runtime/wasm-bpf-rs/src/state.rs:45]: maybe you can wrapper within another struct with arc mutex...etc?

File: runtime/wasm-bpf-rs/src/lib.rs (hunk 2 above)
Comment (Officeyutong) [reply runtime/wasm-bpf-rs/src/lib.rs:232]: According to https://docs.rs/wasmtime/latest/wasmtime/struct.Store.html#method.epoch_deadline_trap, the program can not be recovered when being trapped

File: runtime/wasm-bpf-rs/src/state.rs (hunk 1 above)
Comment (Officeyutong) [reply runtime/wasm-bpf-rs/src/state.rs]: Fixed. this unsafe impl is no longer needed

**Comment by Officeyutong** [comment]
I write a new implementation of `WasiFile`, which provides ability that can read bytes in it when the wasm program running.

File: runtime/wasm-bpf-rs/src/tests/mod.rs [hunk 3]
```diff
@@ -0,0 +1,142 @@
+use crate::tests::pipe::ReadableWritePipe;
+
+use super::*;
+use std::fs::File;
```
Comment (yunwei37) [inline runtime/wasm-bpf-rs/src/tests/mod.rs:4]: move the pipe implement out of test because ecli will needs it.

**Review by yunwei37 — APPROVED** [review]

File: runtime/wasm-bpf-rs/src/tests/mod.rs (hunk 3 above)
Comment (Officeyutong) [reply runtime/wasm-bpf-rs/src/tests/mod.rs:4]: Seems you are reviewing an outdated version of code....

**Pull request merged**